policy_parser.py

```diff
@@ -14,35 +14,107 @@ def __init__(self, msg=None, policy=None):
+        Throws PolicyError if the Policy is invalid or unsupported.
+        """
+        _log.debug("Calculating pod selector")
+        k8s_selector = self.policy["spec"]["podSelector"]
+
+        # PodSelectors only select pods from the Policy's namespace.
+        calico_selectors = ["%s == '%s'" % (K8S_NAMESPACE_LABEL, self.namespace)]
+
+        calico_selectors += self._calculate_selectors(k8s_selector)
+      
+        _log.debug("Selector with %d filters" % len(calico_selectors))
+        return " && ".join(calico_selectors)
 
     def calculate_inbound_rules(self):
         """
-        Takes a NetworkPolicy object from the API and returns a list of
-        Calico Rule objects which should be applied on ingress.
+        Generate Calico Rule objects for this Policy's ingress rules.
+
+        Returns a list of Calico datamodel Rules.
+        Throws PolicyError if the Policy is invalid or unsupported.
         """
         _log.debug("Calculating inbound rules")
         rules = []
 
-        # Iterate through and create the appropriate Calico Rule objects.
-        allow_incomings = self.policy["spec"].get("ingress") or []
-        _log.debug("Found %s ingress rules", len(allow_incomings))
-        for allow_incoming_clause in allow_incomings:
-            # Convert each allow_incoming_clause into one or more
-            # Calico Rule objects.
-            if allow_incoming_clause:
-                # Rule exists - parse it.
-                r = self._allow_incoming_to_rules(allow_incoming_clause)
-            else:
-                # An empty rule means allow all traffic.
-                r = [Rule(action="allow")]
-            rules.extend(r)
+        ingress_rules = self.policy["spec"].get("ingress")
+        if ingress_rules:
+            _log.debug("Got %d ingress rules: translating to Calico format",
+                       len(ingress_rules))
+            for ingress_rule in ingress_rules:
+                _log.debug("Processing ingress rule %s", ingress_rule)
+                if ingress_rule:
+                    # Convert ingress rule into Calico Rules.
+                    _log.debug("Adding rule %s", ingress_rule)
+                    rules.extend(self._allow_incoming_to_rules(ingress_rule))
+                else:
+                    # An empty rule means allow all traffic.
+                    _log.debug("Empty rule => allow all; skipping rest")
+                    rules.append(Rule(action="allow"))
+                    break
 
         _log.debug("Calculated total set of rules: %s", rules)
         return rules
 
+    def _calculate_selectors(self, label_selector, key_format="%s"):
+        """
+        Generate Calico datamodel selectors for a Kubernetes LabelSelector.
+
+        Returns a list of selectors in the Calico datamodel format.
+        Throws PolicyError if the LabelSelector is invalid or unsupported.
+        """
+        # A null LabelSelector matches no objects.
```

Comment:
Not sure this is quite right.

From http://kubernetes.io/docs/user-guide/labels/
> An empty label selector (that is, one with zero requirements) selects every object in the collection.
A null label selector (which is only possible for optional selector fields) selects no objects.


I think this means the following selector should select _all_ endpoints that correspond to pods in this namespace:
```python
{
  ...
  "selector": {
    "matchLabels": { },
    "matchExpressions": [ ]
  }
  ...
}
```

Whereas this selector should select no objects:
```python
{
  "selector": None
}
```

Worth doing a bit more investigation.